tools/ollama.go

```diff
@@ -87,3 +105,155 @@ func (*ollama) Install(logger *log.Logger, installPath string) (err error) {
+	// ollamaProcessStatesLock holds the lock for accessing Ollama process states.
+	ollamaProcessStatesLock sync.Mutex
+)
+
+type ollamaProcessState struct {
+	// url holds the URL of the Ollama process.
+	url string
+	// connectionCount holds the amount of connections.
+	// When the counter hits `0`, the process will be removed.
+	connectionCount uint
+	// cleanup holds functio to clean up the Ollama process.
+	// Can be `nil` in case the process was running already.
+	cleanup func()
+}
+
+// ollamaProcessStateAdd add a connection to the process states.
+func ollamaProcessStateAdd(url string, cleanup func()) {
+	state, ok := ollamaProcessStates[url]
+	if !ok {
+		state = &ollamaProcessState{
+			url: url,
+		}
+		ollamaProcessStates[url] = state
+	}
+	if cleanup != nil {
+		state.cleanup = cleanup // A process can be started afterwards as well, so make sure it is always cleaned up.
+	}
+	state.connectionCount++
+}
+
+// ollamaProcessStateRemove remove a connection from the process states.
+func ollamaProcessStateRemove(url string) {
+	state, ok := ollamaProcessStates[url]
+	if !ok {
+		panic("removed state without being added")
+	}
+
+	state.connectionCount--
+	if state.connectionCount < 0 {
+		panic("removed stated more often than added")
+	} else if state.connectionCount == 0 {
+		if state.cleanup != nil {
+			state.cleanup()
+		}
+		delete(ollamaProcessStates, url)
+	}
+}
+
+// OllamaCheck checks that an Ollama service is running at the given URL.
+func OllamaCheck(logger *log.Logger, url string) (err error) {
+	if err := retry.Do(func() error {
+		response, err := http.Get(url)
+		if err != nil {
+			return pkgerrors.WithStack(err)
+		} else if response.StatusCode != http.StatusOK {
+			return pkgerrors.WithStack(fmt.Errorf("unexpected status code %d", response.StatusCode))
+		}
+
+		bodyBytes, err := io.ReadAll(response.Body)
+		if err != nil {
+			return pkgerrors.WithStack(err)
+		} else if string(bodyBytes) != "Ollama is running" {
+			return pkgerrors.WithStack(err)
+		}
+
+		return nil
+	}, retry.Delay(time.Second), retry.Attempts(3)); err != nil {
+		return pkgerrors.WithStack(pkgerrors.WithMessage(err, "unable to start Ollama server"))
+	}
+
+	return nil
+}
+
+// OllamaStart starts an Ollama service if necessary with the given binary connected to the given URL and returns a shutdown function.
+func OllamaStart(logger *log.Logger, binaryPath string, url string) (shutdown func() (err error), err error) {
+	ollamaProcessStatesLock.Lock()
+	defer ollamaProcessStatesLock.Unlock()
+
+	// Check if the URL has already a running service.
+	if err := OllamaCheck(logger, url); err == nil {
+		logger.Printf("Reusing existing Ollama service on %q", url)
+
+		ollamaProcessStateAdd(url, nil)
+
+		return func() (err error) {
+			ollamaProcessStateRemove(url)
+
+			return nil
+		}, nil
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	logger.Printf("Starting new Ollama service on %q using %s", url, binaryPath)
```

Comment:
The url is never set at the server command. I believe it must be an env variable so our exec util should be adapted to support envs?